Interface model: IModel first implementation

pjanck opened the pull request:
First implementation is there. `IModel` knows about its:
- extent (`oip::BBox`) - needed in Viewport (implemented)
- EPSG code (`std::string`) - needed by me for my PhD (not implemented yet)

Still missing:
- `data` would need an array of `IModel`s which get filled when reading files (currently it is left as it is)
- model deletion/addition should be considered
- UI overview of loaded models (with zoom-to feature)

*edit 20201025*: Fixes #9 closes #108 resolves #149

Commit:
Introduced .Base project, moved bounding box implementation and introduced IModel interface.

Commit:
Merge branch 'development' into InterfaceModel

Commit:
Refactored the IModel interface, added comments.

Commit:
Added iModel interface to PointCloud, current implementation is missing.

Commit:
Moved IfcGeometryModel to a separate .h/.cpp files, added IModel interface.

hechth commented on Base/src/IModel.h:
```diff
@@ -0,0 +1,62 @@
+
+	// change the georeferencing coordinate reference system to the newEPSGcode
+	// function takes three coordinate values and returns three transformed coordinates
+    virtual void transformAllPoints(const std::string& newEPSGcode, std::function<std::tuple<double,double,double> const(double,double,double)>& transf) = 0;
```
Why 3 doubles? I OIP should work with it's own data format for 3d points I assume!

Maybe also too specific for the general model.

pjanck replied:
Well - 3d coordinates are transformed to 3d coordinates. I am open to suggestions, what to include here :)

hechth commented on Base/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+
+# Thirdparty libraries the project depends upon.
+target_link_libraries(OpenInfraPlatform.Base PRIVATE
+#    Qt5::Core
+#    Qt5::OpenGL
+#    Qt5::Xml
+#    Qt5::XmlPatterns
+    BlueFramework.Core
+    BlueFramework.Rasterizer
+#    BlueFramework.Application
+    carve
+#    eigen
+)
+
+#set_target_properties(OpenInfraPlatform.Base 	PROPERTIES FOLDER "OpenInfraPlatform")
```
Remove out-commented stuff if checking in new file!

pjanck replied:
Right. Resolved 30dedf2

hechth commented on Base/src/IModel.h:
```diff
@@ -0,0 +1,62 @@
+    // needed for correct correlation between multiple models
+    // also to determine if a transformation would be needed
+	// the function should return empty string if unknown
+    virtual std::string getEPSGcode() const = 0;
```
Maybe move this to another interface called IGeoReferencable?

pjanck replied:
I can see the idea, however, every model should be georeferencable. 

Proposal: Make a new interface `IGeoReferencable`, that the `IModel` includes, but not implements?

Commit:
Removed commented lines from Base/CMakeList.

Commit:
Added isEmpty() function to IModel.

Commit:
Corrected some exceptions throwing.

TODO: exceptions should probably be introduced in Base project?

Commit:
Added IModel::getSource() function.

Provides means that identify the source of the model. Helps the user/programmer identify, where does the model come from.

Commit:
Goes hand-in-hand with previous commit.

Commit:
Refactored Core::DataManagement::Data a bit.

Commit:
Added an array of models to data class.

Commit:
Introduced PointCloud to oip namespace.

Commit:
Refactored ChangeFlag

Commit:
Changed how the models are loaded in the Data class.

Now multiple models can be successfully loaded one after the other, tested with IFCs. The same should be true for point clouds, however not tested yet.

Commit:
Merge branch 'development' into InterfaceModel

Commit:
Introduced GeorefMetadata to IModel

Commit:
Small changes to BBox implementation.

Commit:
IfcPoint is now represented by a cross, if it is a direct representation

Commit:
Merge branch 'development' into InterfaceModel

Commit:
Merge branch 'development' into InterfaceModel

pjanck commented:
Fixes #9 .

Implemented:
- `data` has an array of models, with getters/setters
- `GeorefMetadata` introduced to handle georeferencing metadata

Still missing:
- [x] dialog/UI element with an overview over the loaded models
- [ ] zoom-to feature (see #228 )
- [ ] loading IFC files of different IFC version (see #213 )

Commit:
Added a UI widget to show which models are loaded.

pjanck commented:
View after the last commit, with these files loaded (and zoomed in a bit):
- tessellated-item.ifc
- basin-faceted-brep.ifc

![grafik](https://user-images.githubusercontent.com/59165496/97229563-db9ff480-17d8-11eb-8a10-9c7a4b7fd7b0.png)

Commit:
Removed unnecessary git file.

Commit:
Now some models' properties get shown in the UI.

pjanck commented:
Upgrade: the UI now shows some moels' properties on the right (same models loaded as above).

![grafik](https://user-images.githubusercontent.com/59165496/97241420-57a63680-17f1-11eb-9f9e-6b4867f48b93.png)

Commit:
Merge branch 'development' into InterfaceModel

Commit:
Merge branch 'development' into InterfaceModel

Commit:
Split IFC geometries handling - individual now.

Commit:
Renamed IfcGeometryModel to IfcModel

not everywhere yet, but it compiles

Commit:
Introduced an oip's base Effect class that accounts for offset

Commit:
Changed all effects to derive from the new base class

Commit:
simplified code a bit

Commit:
Removed georeference from data class

pjanck commented:
With the latest changes, the offsets are now calculated correctly and models rendered in correct spatial relationship to one another.

![grafik](https://user-images.githubusercontent.com/59165496/99372342-4ddb9480-28c0-11eb-8230-b1eba2cebae0.png)

Commit:
Clearing of data model changed

Commit:
Added a bounding box of all models to the UI.

Commit:
Remodeled tree view widget of UI showing the loaded models.
